pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarStreamMetadataProvider.java

```diff
@@ -0,0 +1,159 @@
+   * If offset criteria is smallest, the message id of earliest record in the partition is returned.
+   * If offset criteria is largest, the message id of the largest record in the partition is returned.
+   * throws {@link IllegalArgumentException} if the offset criteria is invalid
+   * throws {@link PulsarClientException} if there was an error from pulsar server on requesting message ids.
+   * @param offsetCriteria offset criteria to fetch {@link StreamPartitionMsgOffset}.
+   *                       Depends on the semantics of the stream e.g. smallest, largest for Kafka
+   * @param timeoutMillis fetch timeout
+   */
+  @Override
+  public StreamPartitionMsgOffset fetchStreamPartitionOffset(@Nonnull OffsetCriteria offsetCriteria,
+      long timeoutMillis) {
+    Preconditions.checkNotNull(offsetCriteria);
+    try {
+      MessageId offset = null;
+      if (offsetCriteria.isLargest()) {
+        _reader.seek(MessageId.latest);
+        if (_reader.hasMessageAvailable()) {
+          offset = _reader.readNext().getMessageId();
+        }
+      } else if (offsetCriteria.isSmallest()) {
+        _reader.seek(MessageId.earliest);
+        if (_reader.hasMessageAvailable()) {
+          offset = _reader.readNext().getMessageId();
+        }
+      } else {
+        throw new IllegalArgumentException("Unknown initial offset value " + offsetCriteria.toString());
+      }
+      return new MessageIdStreamOffset(offset);
+    } catch (PulsarClientException e) {
+      LOGGER.error("Cannot fetch offsets for partition " + _partition + " and topic " + _topic + " and offsetCriteria "
+          + offsetCriteria.toString(), e);
+      return null;
+    }
+  }
+
+  /**
+   * We assume 1:1 mapping from partitionGroupId to partitionId in pulsar. This works because pulsar doesn't have
+   * the concept of child partitions as well as reducing the current number of partition.
+   */
+  @Override
+  public List<PartitionGroupMetadata> computePartitionGroupMetadata(String clientId, StreamConfig streamConfig,
+      List<PartitionGroupConsumptionStatus> partitionGroupConsumptionStatuses, int timeoutMillis)
+      throws TimeoutException, IOException {
+    List<PartitionGroupMetadata> newPartitionGroupMetadataList = new ArrayList<>();
+
+    for (PartitionGroupConsumptionStatus partitionGroupConsumptionStatus : partitionGroupConsumptionStatuses) {
+      newPartitionGroupMetadataList.add(
+          new PartitionGroupMetadata(partitionGroupConsumptionStatus.getPartitionGroupId(),
+              partitionGroupConsumptionStatus.getStartOffset()));
+    }
+
+    try {
+      List<String> partitionedTopicNameList = _pulsarClient.getPartitionsForTopic(_topic).get();
+
+      if (partitionedTopicNameList.size() > partitionGroupConsumptionStatuses.size()) {
+        int newPartitionStartIndex = partitionGroupConsumptionStatuses.size();
+
+        for (int p = newPartitionStartIndex; p < partitionedTopicNameList.size(); p++) {
+
+          Reader reader =
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *reader*. Currently, there are execution paths that do not contain closure statements, for example, when Reader.hasMessageAvailable throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).